libs/backend/src/cast_vote_record_helpers.ts

```diff
@@ -0,0 +1,81 @@
+import { Optional, assert } from '@votingworks/basics';
+import { CVR, ContestId, ContestOptionId, Side } from '@votingworks/types';
+
+/**
+ * Returns the current snapshot of a cast vote record, or undefined if none
+ * exists. If undefined, the cast vote record is invalid.
+ */
+export function getCurrentSnapshot(cvr: CVR.CVR): Optional<CVR.CVRSnapshot> {
+  return cvr.CVRSnapshot.find(
+    (snapshot) => snapshot['@id'] === cvr.CurrentSnapshotId
+  );
+}
+
+/**
+ * Information about a write in entry found in a cast vote record. `text`
+ * indicates the text of the write-in which only applies to machine-marked
+ * ballots. `side` indicates the side of the sheet for the corresponding
+ * ballot image for a hand-written write-in. If `side` and `text` are undefined,
+ * then the corresponding image for the hand-written write-in was not found and
+ * the cast vote record may be invalid.
+ */
+export interface CastVoteRecordWriteIn {
+  contestId: ContestId;
+  optionId: ContestOptionId;
+  side?: Side;
+  text?: string;
+}
+
+/**
+ * Gets the write-in votes from a CDF cast vote record. Asserts the current
+ * snapshot exists.
+ */
+export function getWriteInsFromCastVoteRecord(
+  cvr: CVR.CVR
+): CastVoteRecordWriteIn[] {
+  const currentSnapshot = getCurrentSnapshot(cvr);
+  assert(currentSnapshot);
+
+  const castVoteRecordWriteIns: CastVoteRecordWriteIn[] = [];
+
+  for (const cvrContest of currentSnapshot.CVRContest) {
+    for (const cvrContestSelection of cvrContest.CVRContestSelection) {
+      for (const selectionPosition of cvrContestSelection.SelectionPosition) {
+        const cvrWriteIn = selectionPosition.CVRWriteIn;
+        if (
+          selectionPosition.HasIndication === CVR.IndicationStatus.Yes &&
+          cvrWriteIn
+        ) {
+          // Check for a BMD write-in
+          if (cvrWriteIn.Text) {
```

Comment:
Just curious: How does `cvrWriteIn.Text` indicate that the write-in is a BMD write-in?

Also, very minor, but rather than adding a comment to indicate what this check is doing, we could wrap it in a function to make that clear, e.g.

```ts
function isBmdWriteIn(cvrWriteIn: ...): boolean {
  // Might still want some comment explaining why this indicates that the write-in is a BMD
  // write-in
  return Boolean(cvrWriteIn.Text);
}
```

Reply:
That's just how we're using the CDF specification (and how it's intended to be used). If you vote on paper, we don't have a piece of text we only have an image (unless we were using OCR). If you vote on a BMD, you don't have an image you have a piece of text, basically a write-in at the "transcribed" stage.

The function suggestion is great, incorporated in 540047482afa3cf9ca602d4c0609e522463943fc.